### PR TITLE
Add upgrade HEP

### PR DESCRIPTION
**Pull request by bk201**
**Note**: the upgrade feature is implemented in v1.0.0 and v1.0.1. This belated HEP is added to describe how the upgrade works.

[Rendered](https://github.com/harvester/harvester/blob/147c4b1fd8c885300f24b161990fca12fbc6a9bc/enhancements/20220413-zero-downtime-upgrade.md)

**Related Issue:**

https://github.com/harvester/harvester/issues/1854

### CODE REVIEW

File: enhancements/20220413-zero-downtime-upgrade.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+### Related Issues
+
+- https://github.com/harvester/harvester/issues/1022
+
+## Motivation
+
+### Goals
+
+The user can upgrade a Harvester cluster to a newer version from GUI or API. During the upgrade, VMs in the Harvester have no downtime.
```
Comment (guangbochen) [inline enhancements/20220413-zero-downtime-upgrade.md]:
```suggestion
The user can upgrade a Harvester cluster to a newer version from GUI or API. During the upgrade, there will be no downtime for the VMs in the Harvester.
```

File: enhancements/20220413-zero-downtime-upgrade.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+apiVersion: harvesterhci.io/v1beta1
+kind: Upgrade
+metadata:
+  generateName: hvst-upgrade-
+  namespace: harvester-system
+spec:
+  version: "1.0.1"
+```
+
+The upgrade controller will check if a corresponding `Version` CR exists. If the CR exits, the controller starts the upgrade by using the `Version` CR. In the previous example, the correspinding `Version` CR is:
```
Comment (weihanglo) [inline enhancements/20220413-zero-downtime-upgrade.md]:
```suggestion
The upgrade controller will check if a corresponding `Version` CR exists. If the CR exists, the controller starts the upgrade by using the `Version` CR. In the previous example, the correspinding `Version` CR is:
```

File: enhancements/20220413-zero-downtime-upgrade.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+- The Harvester controller creates a post-drain job with `harvester-upgrade:<new_version>` image to:
+  - Wait for RKE2 runtime to upgrade.
+  - Upgrade the node's OS with `cos` script.
+  - Clean up unused images.
+  - Reboot the node.
+- When the node is back, the Harvester controller notifies Rancher that the post-hook is done. Rancher starts to upgrade the next node.
+
+**Phase 5**: Clean-up
+
+The Harvester controller will delete the upgrade repository VM and its image.
```
Comment (weihanglo) [inline enhancements/20220413-zero-downtime-upgrade.md]: Use simple present tense for consistency?

```suggestion
The Harvester controller deletes the upgrade repository VM and its image.
```

File: enhancements/20220413-zero-downtime-upgrade.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        app: 0.2.5
+    ```
+
+### Upgrade strategy
+
+No.
+
+## Note
+
+- With the current approach, the last-upgraded node contains no VMs (VMs are live-migrated out of the node).
```
Comment (weihanglo) [inline enhancements/20220413-zero-downtime-upgrade.md]: I feel like this is like a limitation or caveat. Can we state it more prominently? Something like

```suggestion
### Limitations

- With this proposed approach, the last-upgraded node contains no VMs (VMs are live-migrated out of the node).
```

Comment (bk201) [reply enhancements/20220413-zero-downtime-upgrade.md]: Sound good!

**Review by weihanglo — APPROVED** [review]

File: enhancements/20220413-zero-downtime-upgrade.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+No. The first implementation is included in the GA release 1.0.0.
+
+## Design
+
+### Implementation Overview
+
+#### Version management
+
+- When a new Harvester version is released, the developers need to edit the response file in the [upgrade responder](https://github.com/harvester/upgrade-responder) to include the new [Version](https://github.com/harvester/upgrade-responder/blob/032365e51c204dfc4aa8793410f5e6a0114c276e/upgraderesponder/service.go#L60). The developer also needs to specify the [`MinUpgradableVersion`](https://github.com/harvester/upgrade-responder/blob/032365e51c204dfc4aa8793410f5e6a0114c276e/upgraderesponder/service.go#L63) of the new version.
```
Comment (guangbochen) [inline enhancements/20220413-zero-downtime-upgrade.md]:
```suggestion
- When a new Harvester version is released, developers need to edit the response file in the [upgrade responder](https://github.com/harvester/upgrade-responder) to include the new [Version](https://github.com/harvester/upgrade-responder/blob/032365e51c204dfc4aa8793410f5e6a0114c276e/upgraderesponder/service.go#L60). Developers also need to specify the [`MinUpgradableVersion`](https://github.com/harvester/upgrade-responder/blob/032365e51c204dfc4aa8793410f5e6a0114c276e/upgraderesponder/service.go#L63) of the new version.
```

File: enhancements/20220413-zero-downtime-upgrade.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+- Harvester backend queries new versions from the upgrade responder hourly.
+- If there is an upgradable version, the backend will query `https://releases.rancher.com/harvester/<version>/version.yaml` to import the `Version` CR. The `Version` CR contains the Harvester ISO file's URL and checksum.
+
+#### Upgrade CR
+
+The user can create an `Upgrade` CR either from Harvester GUI or a manifest file:
```
Comment (guangbochen) [inline enhancements/20220413-zero-downtime-upgrade.md]:
```suggestion
The user can create an `Upgrade` CR either from Harvester GUI or using a manifest file:
```

Comment (bk201) [reply enhancements/20220413-zero-downtime-upgrade.md]: Thanks, all fixed.

**Review by guangbochen — APPROVED** [review]

File: enhancements/20220413-zero-downtime-upgrade.md [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+No. The first implementation is included in the GA release 1.0.0.
+
+## Design
+
+### Implementation Overview
+
+#### Version management
+
+- When a new Harvester version is released, developers need to edit the response file in the [upgrade responder](https://github.com/harvester/upgrade-responder) to include the new [Version](https://github.com/harvester/upgrade-responder/blob/032365e51c204dfc4aa8793410f5e6a0114c276e/upgraderesponder/service.go#L60). Developers also need to specify the [`MinUpgradableVersion`](https://github.com/harvester/upgrade-responder/blob/032365e51c204dfc4aa8793410f5e6a0114c276e/upgraderesponder/service.go#L63) of the new version.
+- Harvester backend queries new versions from the upgrade responder hourly.
```
Comment (johnliu55tw) [inline enhancements/20220413-zero-downtime-upgrade.md:64]: If Harvester is deployed in an air-gapped environment, can users still receive the latest available upgrades? Or they should manually add versions CR?

Comment (bk201) [reply enhancements/20220413-zero-downtime-upgrade.md:64]: > can users still receive the latest available upgrades?

No, from GUI, we have a hidden page to let users upload the ISO and the frontend will create the CR.
Of course, it's also possible to create CR manually.

Comment (bk201) [reply enhancements/20220413-zero-downtime-upgrade.md:64]: @johnliu55tw I added the air-gapped scenario: https://github.com/harvester/harvester/pull/2146/commits/9a68a3a0a9754caa1bd51c5907783a975f8e3e63#diff-4024911e730bfb92019ae2cf3868136ba568e013b9d1d41fc09bdf27b86c6b37R41-R47

**Review by johnliu55tw — APPROVED** [review]

**Pull request merged**